File: programmers/pjkfckr/H-Index.py
"""

발표한 논문 n편중 h번 이상 인용된 논문이 h편 이상이고 나머지 논문이 h번 이하 인용되었다면 h의 최대값이 return

인용횟수를 담은 배열 citations

"""



def solution(citations):
    n = len(citations)
    reversed_sort = sorted(citations, reverse=True)

    # 정렬된 배열을 순회하면서 각 원소의 값과 현재 인덱스(+1)를 비교합니다.
    # 현재 인덱스 + 1이 현재 원소의 값보다 크거나 같아지는 지점이 H-Index가 될 수 있습니다.
    for i in range(n):
        if i + 1 > reversed_sort[i]:
            return i

    return len(citations)





citations = [3, 0, 6, 1, 5]

print(solution(citations))

File: leetcode/0chnxxx/H-Index.kt
/**
 * Given an array of integers citations where citations[i] is the number of citations a researcher received for their ith paper, return the researcher's h-index.
 *
 * According to the definition of h-index on Wikipedia:
 * The h-index is defined as the maximum value of h such that the given researcher has published at least h papers that have each been cited at least h times.
 *
 * Constraints:
 * n == citations.length
 * 1 <= n <= 5000
 * 0 <= citations[i] <= 1000
 */

fun main() {
    val citations = intArrayOf(3, 0, 6, 1, 5)

    val result = Solution().hIndex(citations)

    println(result)
}

class Solution {
    // 시간 복잡도 : O(N)
    // 공간 복잡도 : O(N)
    fun hIndex(citations: IntArray): Int {
        val n = citations.size
        val count = IntArray(n + 1)

        for (c in citations) {
            if (c >= n) {
                count[n]++
            } else {
                count[c]++
            }
        }

        var papers = 0

        for (i in n downTo 0) {
            papers += count[i]

            if (papers >= i) {
                return i
            }
        }

        return 0
    }

//    // 시간 복잡도 : O(N log N)
//    // 공간 복잡도 : O(N)
//    fun hIndex(citations: IntArray): Int {
//        val sorted = citations.sortedDescending()
//        var h = 0
//
//        for (i in sorted.indices) {
//            if (sorted[i] >= i + 1) {
//                h = i + 1
//            } else {
//                break
//            }
//        }
//
//        return h
//    }
}